### ISSUE TITLE
[bug] Race condition in project share?

### ISSUE THREAD

**Issue by jcmcken** [gitlabhq/terraform-provider-gitlab#109]
### Terraform Version

```
Terraform v0.11.11
+ provider.gitlab v1.2.0
```

### Affected Resource(s)

- ``gitlab_project``

### Terraform Configuration Files
```hcl
provider "gitlab" {
  base_url = "https://gitlab.example.com"
  token    = "${var.gitlab_token}"
}

variable "gitlab_token" {
  type = "string"
}

variable "workspace_name" {
  type = "string"
}

resource "gitlab_project" "project" {
  name = "${var.workspace_name}"

  namespace_id                                     = 1322
  visibility_level                                 = "private"
  approvals_before_merge                           = 1
  only_allow_merge_if_pipeline_succeeds            = true
  only_allow_merge_if_all_discussions_are_resolved = true

  shared_with_groups = [
    {
      group_id           = 9
      group_access_level = "developer"
    },
  ]
}
```

### Expected Behavior

Provider creates project and shares it with no errors.

### Actual Behavior

Sometimes this succeeds without errors,, but sometimes it doesn't.

Every so often, Terraform produces an error:

```
* gitlab_project.project: POST https://gitlab.example.com/api/v4/projects/2947/share: 404 {message: Not Found}
```

If I try to re-apply, the provider acts as if the project has not been created (state is wrong), but then gives this error:

```
* gitlab_project.project: POST https://gitlab.example.com/api/v4/projects: 400 {message: {limit_reached: []}, {name: [has already been taken]}, {path: [has already been taken]}}
```

### Steps to Reproduce

1. Run ``terraform apply`` with the appropriate variables multiple times (perhaps using a GitLab instance with moderate utilization and 100s to 1000s of users).

### Important Factoids

We're using the latest GitLab 11.7.x Enterprise with 'Ultimate'-level license on a moderately sized GitLab installation (hundreds of users with thousands of projects).

### References

Looking [at the code](https://github.com/terraform-providers/terraform-provider-gitlab/blob/master/gitlab/resource_gitlab_project.go#L194), the ``gitlab_project`` resource only validates that the 'create project' call returns successfully. Just spit-balling, but is this request completely synchronous? Perhaps there needs to be a subsequent request to validate the project was created successfully before performing the share operation?

**Comment by jcmcken** [comment]
Nevermind, I don't think there's a race condition. I was accidentally using different users to run the apply, which is why it sometimes succeeded. The real issue is that one user I tried to run it with didn't have permissions to the group it was trying to share the project with. 

It seems weird that you get an HTTP 404 in this case, which is really confusing. Maybe it's just GitLab's way of preventing you from enumerating groups. Perhaps the error message could be updated to reflect this possibility?

**Comment by zhilyaev** [comment]
I created new **clean** gitlab  server (v. 13.2.3). And after terraform apply i catched the same issue. hmm...

**Comment by armsnyder** [comment]
@jcmcken It makes sense that it would be a permissions issue, since GitLab uses 404 for permission denied. The issue you saw with the re-apply causing a 400 should have been fixed in f1610fcb0efb2e59cad07713d5a3f1481e859308.

@zhilyaev Can you share your Terraform config and repro steps? I was unable to reproduce this myself, running apply and destroy multiple times with this config:

```terraform
resource "gitlab_group" "group" {
  name = "my-group"
  path = "my-group"
}

resource "gitlab_project" "project" {
  name = "my-project"
  visibility_level = "private"
  shared_with_groups {
    group_id = gitlab_group.group.id
    group_access_level = "developer"
  }
}
```

**Comment by armsnyder** [comment]
Closing as cannot reproduce and lacking of a response.
